### PR TITLE
upgrade: fetch all formulae before installing

### PR DESCRIPTION
**Pull request by FnControlOption**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/upgrade.rb [hunk 1]
```diff
@@ -41,26 +41,29 @@ def upgrade_formulae(
...
+          quiet:                      quiet,
+          verbose:                    verbose,
+        )
+      rescue UnsatisfiedRequirements, DownloadError => e
+        ofail "#{formula}: #{e}"
+        nil
+      end.compact
+
+      formula_installers.each do |fi|
+        Migrator.migrate_if_needed(fi.formula, force: force)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/upgrade.rb:64]: This should probably be done before `fetch`

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good so far! Nothing to do here but in future reference you can just force push to an existing PR instead of opening a new one.

File: Library/Homebrew/upgrade.rb [hunk 2]
```diff
@@ -82,7 +87,7 @@ def print_upgrade_message(formula, fi_options)
       EOS
     end
 
-    def upgrade_formula(
+    def fetch_formula_installer(
```
Comment (MikeMcQuaid) [inline Library/Homebrew/upgrade.rb:90]:
```suggestion
    def fetched_formula_installer(
```
or
```suggestion
    def fetch_and_return_formula_installer(
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good! One naming thought.

File: Library/Homebrew/upgrade.rb [hunk 3]
```diff
@@ -44,7 +44,7 @@ def upgrade_formulae(
       formula_installers = formulae_to_install.map do |formula|
         Migrator.migrate_if_needed(formula, force: force)
         begin
-          fetch_formula_installer(
+          fi = create_formula_installer(
```
Comment (MikeMcQuaid) [inline Library/Homebrew/upgrade.rb:47]:
```suggestion
          fi = create_and_fetch_formula_installer(
```

File: Library/Homebrew/upgrade.rb [hunk 4]
```diff
@@ -87,7 +89,7 @@ def print_upgrade_message(formula, fi_options)
       EOS
     end
 
-    def fetch_formula_installer(
+    def create_formula_installer(
```
Comment (MikeMcQuaid) [inline Library/Homebrew/upgrade.rb:92]:
```suggestion
    def create_and_fetch_formula_installer(
```

File: Library/Homebrew/upgrade.rb [hunk 5]
```diff
@@ -134,11 +136,8 @@ def fetch_formula_installer(
           verbose:                    verbose,
         }.compact,
       )
-
-      fi.fetch
-      fi
     end
-    private_class_method :fetch_formula_installer
+    private_class_method :create_formula_installer
```
Comment (MikeMcQuaid) [inline Library/Homebrew/upgrade.rb:140]:
```suggestion
    private_class_method :create_and_fetch_formula_installer
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @FnControlOption!

**Comment by FnControlOption** [comment]
Why rename to `create_and_fetch_formula_installer` when `fi.fetch` happens outside the method?

**Comment by MikeMcQuaid** [comment]
@FnControlOption I misread your most recent changes, sorry 🤦🏻. I can open a PR to revert or you can, what would you rather?

**Pull request merged**